ReadMe.md
[![N|Solid](https://raw.githubusercontent.com/Xonshiz/comic-dl/master/Images/Icon.png)](https://github.com/Xonshiz/comic-dl)
# Comic-DL
[![GitHub release](https://img.shields.io/github/release/xonshiz/comic-dl.svg?style=flat-square)](https://github.com/xonshiz/comic-dl/releases/latest) |
[![Github All Releases](https://img.shields.io/github/downloads/xonshiz/comic-dl/total.svg?style=flat-square)](https://github.com/xonshiz/comic-dl/releases) |
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/xonshiz)

| Service           |                                                                                                       Status                                                                                                        |
|:------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
| PyPI Distribution |                [![PyPI](https://github.com/Xonshiz/comic-dl/actions/workflows/python-package.yml/badge.svg?branch=master)](https://github.com/Xonshiz/comic-dl/actions/workflows/python-package.yml)                |  |
| Releases Build    | [![Building & Creating Distributables](https://github.com/Xonshiz/comic-dl/actions/workflows/python-release.yml/badge.svg?branch=master)](https://github.com/Xonshiz/comic-dl/actions/workflows/python-release.yml) |  |
| Docker Build      |                [![GitHub Workflow Status](https://img.shields.io/github/workflow/status/xonshiz/comic-dl/Docker?label=docker%20build)](https://github.com/xonshiz/comic-dl/pkgs/container/comic-dl/)                |  |
| Readthedocs Build |                                 [![Documentation Status](https://readthedocs.org/projects/comic-dl/badge/?version=latest)](https://comic-dl.readthedocs.io/en/latest/?badge=latest)                                 |

Comic-dl is a command line tool to download Comics and Manga from various Manga and Comic sites easily. You can search Manga from this tool as well. Idea from [youtube-dl](https://github.com/rg3/youtube-dl).

If you're looking for an application, or a UI for this, please move to : [CoManga](https://github.com/Xonshiz/CoManga)

> Don't overuse this script. Support the developers of those websites by disabling your adblock on their site.
>Advertisements pay for the website servers.

## Table of Content

* [Supported Sites](https://github.com/Xonshiz/comic-dl/blob/master/Supported_Sites.md)
* [Dependencies Installation](#dependencies-installation)
    * [Linux/Debian](#linuxdebian-)
    * [Windows](#windows-)
    * [Mac OS X](#mac-os-x-)
* [Installation](#installation)
    * [Docker](#docker)
* [Python Support](#python-support)
* [Windows Binary](#windows-binary)
* [List of Arguments](#list-of-arguments)
* [Language Codes](#language-codes)
* [Usage](#usage)
    * [Windows](#windows)
    * [Linux/Debian](#linuxdebian)
* [Auto Download](#auto-download)
* [Features](#features)
* [Changelog](https://github.com/Xonshiz/comic-dl/blob/master/Changelog.md)
* [Opening An Issue/Requesting A Site](#opening-an-issuerequesting-a-site)
    * [Reporting Issues](#reporting-issues)
    * [Suggesting A Feature](#suggesting-a-feature)
* [Contributors](https://github.com/Xonshiz/comic-dl/blob/master/Contributors.md)
* [Notes](#notes)
* [Donations](#donations)

## Supported Websites
You can check the list of supported websites [**`HERE`**](https://github.com/Xonshiz/comic-dl/blob/master/Supported_Sites.md).

## Dependencies Installation
This script can run on multiple Operating Systems. You need `Node.js` in your system's path for this script to work (You need this on each and every Operating System, even on WINDOWS :/). Download the `Node.Js` from [Node.js official website](https://nodejs.org/en/). Doesn't matter which operating system you're on, this is a must. Follow the instructions mentioned below, according to your OS.

### Linux/Debian :
Since most (if not all) Linux/Debian OS come with python pre-installed, you don't have to install python manually. Make sure you're using python >= 2.7.x though.

We need `pip` to install any external dependency(ies). So, open any terminal and type in `pip list` and if it shows some data, then it is fine. But, if it shows error, like `pip not found` or something along this line, then you need to install `pip`. Just type this command in terminal :

`sudo apt-get install python-pip`

If you're on Fedora, CentOS/RHEL, openSUSE, Arch Linux, then you simply need to follow [`THIS TUTORIAL`](https://packaging.python.org/install_requirements_linux/) to install `pip`.

If this still doesn't work, then you'll manually need to install pip. Doing so is an easy one time job, and you can follow   [`THIS TUTORIAL`](https://pip.pypa.io/en/stable/installing/) to do so.

### Windows :
If you're on Windows, then it is recommended to download the [`windows binary`](https://github.com/Xonshiz/comic-dl#windows-binary) for this script. If you use the windows binary, you don't need to install anything, except Node.js. But, if for some weird reason you want to use Python script instead, then follow these steps :

* Install Python > 2.7.6. Download the desired installer from [here](https://www.python.org/downloads/).
* [Add it in the system path](http://superuser.com/questions/143119/how-to-add-python-to-the-windows-path) (if not already added).
* If you're using python >2.7.9, you don't need to install `PIP`. However, if you don't have pip installed and added in windows path, then do so by following [this little tutorial](http://stackoverflow.com/a/12476379).
* Download [this `text`](https://github.com/Xonshiz/comic-dl/blob/master/requirements.txt) file and put it in some directory/folder.
* Open Command Prompt and browse to the directory where you downloaded your requirements.txt file and run this command :
```
pip install -r requirements.txt
```
* It should install the required external libraries.

Now, install Node.Js as well and make sure it's in your path.

Well, if everything came up good without any error(s), then you're good to go!

### Mac OS X :
Mac OS X users will have to fetch their version of `Python` and `Pip`.
* Python installation guide : http://docs.python-guide.org/en/latest/starting/install/osx/
* Pip installation guide : http://stackoverflow.com/questions/17271319/installing-pip-on-mac-os-x

After downloading and installing these, you need to add PIP & Python in your path. Follow [`THIS LITTLE GUIDE`](http://www.pyladies.com/blog/Get-Your-Mac-Ready-for-Python-Programming/) to install both, Python & pip successfully.

## Installation
After installing and setting up all the dependencies in your Operating System, you're good to go and use this script.
The instructions for all the OS would remain same. Download [`THIS REPOSITORY`](https://github.com/Xonshiz/comic-dl/archive/master.zip) and put it somewhere in your system. Move over to the `comic_dl` folder.

**Windows users**, it's better to not place it places where it requires administrator privileges. Good example would be `C:\Windows`. This goes for both, the Python script and the windows binary file (.exe).

**Linux/Debian** users make sure that this script is executable. Just run this command, if you run into problem(s) :

`chmod +x cli.py`

and then, execute with this :

`./cli.py`


### Docker

With docker, you can get the whole dependencies enclosed in a container and use the `comic_dl` from your system.

You need an up and running Docker client running, follow the [Docker Documentation](https://docs.docker.com/install/).

[Docker images are available here](https://github.com/Xonshiz/comic-dl/pkgs/container/comic-dl/) 

Define a handy alias on your system with some docker tricks. This mounts the local directory under `/directory` in the container. This works on *NIX systems, and also under Windows Linux subsystem.

You can change the value of PGID and PUID with the value for the user needed in your download directory.

```bash
alias comic_dl="docker run -it --rm -e PGID=$(id -g) -e PUID=$(id -u) -v $(pwd):/directory:rw -w /directory ghcr.io/xonshiz/comic-dl:latest comic_dl -dd /directory"
```

Run it on your system. This actually starts a container on request and stop&delete it when finished.

```bash
usage: comicdl [-h] [--version] [-s SORTING] [-a] [-c]
               [-dd DOWNLOAD_DIRECTORY] [-rn RANGE] [--convert CONVERT]
               [--keep KEEP] [--quality QUALITY] [-i INPUT] [--comic]
               [-comic-search SEARCH_COMIC] [-comic-info COMIC_INFO]
               [--update UPDATE] [--print-index]
               [-ml MANGA_LANGUAGE] [-sc SKIP_CACHE] 
               [-p PASSWORD] [-u USERNAME] [-v]
[...]
```

## Python Support
This script supports python 3. You can run this on python 2 as well, as long as you have all the dependencies installed.

## Windows Binary
It is recommended that Windows users use this binary to save both, your head and time from installing all the dependencies.

You need to download and install [Node.js](https://nodejs.org/en/) and make sure it is in your Windows path (watch out for the tick box during install).

If you already have it, then you can download this binary and start using the script right off the bat :
* `Binary (x86)` : [Click Here](https://github.com/Xonshiz/comic-dl/releases/latest)

**`NOTE:`** This is a `COMMAND LINE TOOL` and will `NOT` work when you double-click it. So, download this `.exe` file and put it in some folder (Not in windows or some other restricted folder) and then hold down `SHIFT KEY` and right click anywhere on the screen and select "`Open Command Window here`". It'll open a CMD window for you. Now, you need to execute it and pass arguments to it like:

```
comic_dl.exe -i "https://readcomicsonline.ru/comic/irredeemable-omnibus-2012"
```

Read which argument does what in [List of Arguments](#list-of-arguments) section.


## List of Arguments
Currently, the script supports these arguments :
```
-h, --help                             Prints the basic help menu of the script and exits.
-i,--input                             Defines the input link to the comic/manga.
--print-index                          Prints the range index for links in the input URL
-V,--version                           Prints the VERSION and exits.
-u,--username                          Indicates username for a website.
-p,--password                          Indicates password for a website.
-v,--verbose                           Enables Verbose logging.
--sorting							   Sorts the download order.(VALUES = asc, ascending,old,new,desc,descending,latest,new)
-a, --auto                             Download new chapters automatically (needs config file!)
-c, --config                           Generates config file for autodownload function
-dd,--download-directory               Specifies custom download location for the comics/manga.
-rn,--range                            Selects the range of Chapters to download (Default = All) [ Ex : --range 1-10 (This will download first 10 episodes of a series)]
--convert						       Tells the script to convert the downloaded Images to PDF or anything else. (Supported Values : pdf, cbz) (Default : No) [By default, script will not convert anything.]
--keep   							   Tells the script whether to keep the files after conversion or not. (Supported : No, False) (Default : Yes/True) [By default, images will be kept even after conversion.]
--quality   						   Tells the script about the image quality you want to download. (Supported Values : low/bad/worst/mobile/cancer) [By default, images will be downloaded in Highest Quality Available. No need to provide any option.]
-ml, --manga-language                  Selects the language for manga. 0 is English (Default) and 1 is Italian.
-sc, --skip-cache                      Forces to skip cache checking.
--comic                                Add this after -i if you are inputting a comic id or the EXACT comic name.
                                       [ Ex : -i "Deadpool Classic" --comic ]
-comic-search, --search-comic          Searches for a comic through the scraped data from ReadComicOnline.li
                                       [ Ex : -comic-search "Deadpool" ]
-comic-info, --comic-info              Lists all the information about the given comic (argument can be either comic id or the exact comic name).
                                       [ Ex : -comic-info "Deadpool Classic" ] or [ Ex : -comic-info 3865 ]
--update                               Updates the comic database for the given argument.
                                       [ Ex: --update "Deadpool Classic" ] or [ Ex: --update "https://readcomiconline.li/Comic/Deadpool-Classic" ]
-cookie, --cookie                      Passes a cookie to be used throughout the session.
```

## Language Codes:

Changelog.md
# Changelog

- Site support for bato.to [2016.11.15]
- Bug fix in comic-dl.py and added docs for readthedocs & updated readme [2016.11.20]
- Site support for kissmanga.com [2016.11.22]
- Addition of universal `downloader package` for easy maintainance [2016.11.22]
- Mangafox downloading same images fixed [2016.11.22]
- Argument priority updated [2016.11.22]
- Site support for comic.naver.com [2016.11.26]
- Support for Python 3 [2016.11.26]
- Removed Kissmanga PhantomJS dependency [2016.12.23]
- Support for Python 2 and 3 has been merged [2016.12.28]
- Updated PhantomJS dependency in [supported sites](https://github.com/Xonshiz/comic-dl/blob/master/Supported_Sites.md) [2016.12.28]
- Added Icon (Logo) [2016.12.28]
- Added a YouTube Tutorial for the script [2016.12.30]
- Site support for readcomiconlin.to [2017.01.02]
- Added `Verbose Logging` [2017.01.22]
- Fixed chapter count error in Kissmanga [2017.01.22]
- Fixed #4 [2017.02.16]
- Optimized Imports [2017.02.16]
- Site support for mangahere.co [2017.02.23]
- Added `Sorting Order` a.k.a `Download Order` [2017.02.23]
- Fixed #5 [2017.02.28]
- Site support for raw.senmanga.com [2017.03.06]
- Site support for omgbeaupeep.com [2017.05.07]
- Fix for image links in omgbeaupeep.com [2017.05.08]
- Re-wrote whole script to implement classes [2017.06.12]
- Better Data Flow and faster than before [2017.06.12]
- Better Log generation [2017.06.12]
- Requirement for Selenium and PhantomJS completely removed [2017.06.12]
- Fix for #9 [2017.06.12]
- Fix for #16 [2017.06.12]
- Fix for #18 [2017.06.12]
- Fix for #21 [2017.06.12]
- Downloader now prints file size, internet speed [2017.06.12]
- Added Support for [ac.qq.com](http://ac.qq.com) [2017.06.15]
- Fix for #23 [2017.06.30]
- Custom Download Location Support Added (#20) [2017.06.30]
- Support to download only selected chapters (range) [2017.07.06]
- Fix for #32 [2017.07.07]
- Support for converting downloaded images into pdf #17 [2017.07.07]
- Fix for #33 [2017.07.07]
- Fix for #34 [2017.07.24]
- Fix for #36 [2017.07.24]
- Fix for #37 [2017.07.24]
- Fix for #30 [2017.08.23]
- Fix for #30 (mangahere.co) [2017.08.23]
- Added support for CBZ file creation (#28) [2017.08.24]
- Added Support for [STRIP-UTOPIJA](http://striputopija.blogspot.in/) (#19) [2017.09.20]
- Fixed Major Bug in ReadComicOnline.to "Full Series" download section [2017.09.21]
- Added support for choosing image quality in ReadComicOnline.to (#42) [2017.09.21]
- Fix for #47 [2017.09.24]
- Fix for #[42](https://github.com/Xonshiz/comic-dl/issues/42#issuecomment-331693291) [2017.09.24]
- Updated ReadMe.md do add missing Node.js step for Windows binary users [2017.09.24] - @DavidFarrell [#48]
- Fix For #51 [2017.11.15] - @dsanchezseco [#52]
- Fix for #53 [2017.11.15] - @dsanchezseco [#54]
- Fix conversion filepath of cbz format [2017.11.15] - @dsanchezseco [#55]
- Fix For #50 [2017.11.26] - @Tyroc [#56]
- Added Support for [MangaReader.net](http://mangareader.net/) (#43) [2017.12.22]
- Added Support for [readcomicbooksonline.net](http://readcomicbooksonline.net/) (#57) [2017.12.26]
- Fix For #60 [2017.12.26]
- Fix For #61 [2017.12.26]
- Added Support for [readcomics.website](http://www.readcomics.website/) (#63) [2017.12.27]
- Fix For #64 [2017.12.27]
- Fix For #66 [2017.12.27]
- Added MangaFox.la (mangafox.me redirects to .la now) [2017.12.27]
- Added Feature to Search and Download Manga from MangaEden [2017.12.28]
- Fix for #64 [2017.12.28]
- Fix for [MangaRock.com](https://mangarock.com) #62 [2017.12.29]
- Added Support for [Bato.to](http://bato.to/) [2017.12.31] [No PhantomJS Needed Now]
- Fix for #69 [2017.12.31]
- Fix for 72 [2018.01.09]
- Removed un-necessary code [2018.01.09]
- Fix for 73 [2018.01.10]
- Fix for 74 [2018.01.10]
- By default, script will not delete the file [2018.01.19]
- Fix for #79 [2018.01.19]
- Fix for #80 [2018.01.19]
- Fix for #90 [2018.01.20]
- Support for Auto Download Included [2018.01.21] (@dsanchezseco)
- Error Log creation to be done in the desired download directory now. [2018.01.22]
- Fix for #89 [2018.01.22]
- Fix for #93 [2018.01.28]
- Fix for #94 [2018.01.28]
- Fix for #99 [2018.02.01] (@dsanchezseco)
- Fix for #103 [2018.02.01] (@dsanchezseco)
- Fix for #105 [2018.02.03]
- Fix for #102 [2018.02.03]
- Fix for #98 [2018.02.03]
- Code Formatting [2018.02.03]
- Added Support for [hqbr.com.br](https://hqbr.com.br/home) [2018.03.15]
- Added Support for [comicextra.com](http://www.comicextra.com/) [Fix for #116] [2018.03.23]
- Fix for #119 [2018.04.10]
- Fix for #120 [2018.04.10]
- Fix for #121 [2018.04.10]
- Added Support for [readcomics.io](https://www.readcomics.io/) [Fix for #117] [2018.04.11]
- Fix for #124 [2018.04.29]
- Fix for #18 [2018.04.29]
- PR Merge for #126 (#127)  [2018.05.07]
- Fix for #128 [2018.05.07]
- PR Merge for #131 (#136)  [2018.05.12]
- PR Merge for #130 (#129)  [2018.05.16]
- PR Merge for #135 (#133)  [2018.05.20]
- PR Merge for #137 (#129 and #136)  [2018.05.21]
- PR Merge for #141
- PR Merge for #142 
- PR Merge for #144 
- PR Merge for #149 
- PR Merge for #151 
- PR Merge for #157 
- PR Merge for #160 
- PR Merge for #161 
- PR Merge for #162 
- PR Merge for #167 
- PR Merge for #172  [2018.09.24]
- PR Merge for #189   [2018.11.05]
- PR Merge for #199   [2018.11.27]
- All PRs Merged [2019.05.26]
- All PRs Merged [2021.06.13]
- Fixing import levels [2021.09.04]
- Added a 5 second wait while downloading whole chapter from RCO as suggested in #261 [2021.09.05]
- Fixed #220 [2021.09.05]
- Added support for readcomicsonline.ru [Fix for #290] [2021.09.05]
- Updated ReadMe to include "execution" on windows [2021.09.05]
- Removed setup2.py file [2021.09.05]
- Checking for existing CBZ/PDF files before downloading them again [Fix for #247] [2021.09.05]
- Fix for chapter download at readmanganato
- Added support for webtoons.com (No audio download yet) [Fix for #284] [2021.09.05.1]
- Fix for #299 [2022.04.16]
- Added support for TMOFans/lectortmo.com [Fix for #231] [2022.04.17]

comic_dl/honcho.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

try:
    from urllib.parse import urlparse
except ImportError:
    from urlparse import urlparse
import logging
from .sites import foolSlide
from .sites import readcomicOnlineli
from .sites import comicNaver
from .sites import mangaHere
from .sites import rawSenManga
from .sites import mangaFox
from .sites import omgBeauPeep
from .sites import mangaReader
from .sites import mangaEden
from .sites import acQQ
from .sites import stripUtopia
from .sites import readComicBooksOnline
from .sites import readComicsWebsite
from .sites import batoto
from .sites import hqbr
from .sites import comicextra
from .sites import readComicsIO
from .sites import japscan
from .sites import manganelo
from .sites import webtoons
from .sites import lectortmo
from .sites import mangatoonMobi
from .sites import mangadex
from .sites import quiremanhua


class Honcho(object):
    def comic_language_resolver(self, language_code):
        # Will return the Language Name corresponding to the language code.
        language_dict = {
            '0': 'English',
            '1': 'Italian',
            '2': 'Spanish',
            '3': 'French',
            '4': 'German',
            '5': 'Portuguese',
            '6': 'Turkish',
            '7': 'Indonesian',
            '8': 'Greek',
            '9': 'Filipino',
            '10': 'Polish',
            '11': 'Thai',
            '12': 'Malay',
            '13 ': 'Hungarian',
            '14': 'Romanian',
            '15': ' Arabic',
            '16': 'Hebrew',
            '17': 'Russian',
            '18': 'Vietnamese',
            '19': 'Dutch',
            '20': 'Bengali',
            '21': 'Persian',
            '22': 'Czech',
            '23': 'Brazilian',
            '24': 'Bulgarian',
            '25': 'Danish',
            '26': 'Esperanto',
            '27': 'Swedish',
            '28': 'Lithuanian',
            '29': 'Other'
        }
        return language_dict[language_code]

    def checker(self, comic_url, download_directory, chapter_range, **kwargs):

        user_name = kwargs.get("username")
        password = kwargs.get("password")
        current_directory = kwargs.get("current_directory")
        log_flag = kwargs.get("logger")
        sorting = kwargs.get("sorting_order")
        comic_language = kwargs.get("comic_language")
        print_index = kwargs.get("print_index")
        manual_cookies = kwargs.get("cookie", None)

        if log_flag is True:
            logging.basicConfig(format='%(levelname)s: %(message)s', filename="Error Log.log", level=logging.DEBUG)
            logging.debug("Comic Url : %s" % comic_url)

        domain = urlparse(comic_url).netloc
        logging.debug("Selected Domain : %s" % domain)

        # Remove the "/" from ending to make checking URL for Full Series or Single Chapter easier.
        if comic_url[-1] == "/":
            comic_url = comic_url[:-1]

        if domain in ["yomanga.co", "gomanga.co"]:
            foolSlide.FoolSlide(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                keep_files=kwargs.get("keep_files"))
            return 0
        elif domain in ["www.readcomiconline.li", "readcomiconline.li", "www.readcomicsonline.ru",
                        "readcomicsonline.ru"]:
            readcomicOnlineli.ReadComicOnlineLi(manga_url=comic_url, logger=logging,
                                                current_directory=current_directory, sorting_order=sorting,
                                                log_flag=log_flag, download_directory=download_directory,
                                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                                keep_files=kwargs.get("keep_files"),
                                                image_quality=kwargs.get("image_quality"),
                                                print_index=print_index, manual_cookies=manual_cookies)
            return 0
        elif domain in ["www.comic.naver.com", "comic.naver.com"]:
            comicNaver.ComicNaver(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                  sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                  chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                  keep_files=kwargs.get("keep_files"),
                                  print_index=print_index)
            return 0
        elif domain in ["www.mangahere.co", "mangahere.co", "www.mangahere.cc", "mangahere.cc"]:
            mangaHere.MangaHere(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                keep_files=kwargs.get("keep_files"),
                                print_index=print_index)
            return 0
        elif domain in ["www.raw.senmanga.com", "raw.senmanga.com"]:
            rawSenManga.RawSenaManga(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                     sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                     chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                     keep_files=kwargs.get("keep_files"),
                                     print_index=print_index)
            return 0
        elif domain in ["www.mangafox.me", "mangafox.me", "www.mangafox.la", "mangafox.la", "www.fanfox.net",
                        "fanfox.net"]:
            mangaFox.MangaFox(manga_url=comic_url, logger=logging, current_directory=current_directory,
                              sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                              chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                              keep_files=kwargs.get("keep_files"),
                              print_index=print_index)
            return 0
        elif domain in ["www.omgbeaupeep.com", "omgbeaupeep.com", "www.otakusmash.com", "otakusmash.com"]:
            omgBeauPeep.OmgBeauPeep(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                    sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                    chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                    keep_files=kwargs.get("keep_files"),
                                    print_index=print_index)
            return 0
        #  TODO KO --print-index -i http://ac.qq.com/Comic/comicInfo/id/547059?trace_id=907_27.156.162.231_1539265645  broken?
        elif domain in ["www.ac.qq.com", "ac.qq.com"]:
            acQQ.AcQq(manga_url=comic_url, logger=logging, current_directory=current_directory,
                      sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                      chapter_range=chapter_range,
                      print_index=print_index)
            return 0
        elif domain in ["www.striputopija.blogspot.in", "striputopija.blogspot.in", "www.striputopija.blogspot.com",
                        "striputopija.blogspot.com"]:
            stripUtopia.StripUtopia(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                    sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                    chapter_range=chapter_range,
                                    print_index=print_index)
            return 0
        elif domain in ["www.mangareader.net", "mangareader.net"]:
            mangaReader.MangaReader(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                    sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                    chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                    keep_files=kwargs.get("keep_files"),
                                    print_index=print_index)
            return 0
        elif domain in ["www.readcomicbooksonline.net", "readcomicbooksonline.net", "www.readcomicbooksonline.org",
                        "readcomicbooksonline.org"]:
            readComicBooksOnline.ReadComicBooksOnline(manga_url=comic_url, logger=logging,
                                                      current_directory=current_directory, sorting_order=sorting,
                                                      log_flag=log_flag, download_directory=download_directory,
                                                      chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                                      keep_files=kwargs.get("keep_files"),
                                                      print_index=print_index)
            return 0
        #  TODO KO seems broken
        elif domain in ["www.readcomics.website", "readcomics.website"]:
            readComicsWebsite.ReadComicsWebsite(manga_url=comic_url, logger=logging,
                                                current_directory=current_directory, sorting_order=sorting,
                                                log_flag=log_flag, download_directory=download_directory,
                                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                                keep_files=kwargs.get("keep_files"),
                                                print_index=print_index)
            return 0
        elif domain in ["www.japscan.to"]:
            japscan.Japscan(manga_url=comic_url, logger=logging, current_directory=current_directory,
                            sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                            chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                            keep_files=kwargs.get("keep_files"),
                            print_index=print_index)
            return 0
        elif domain in ["www.hqbr.com.br", "hqbr.com.br"]:
            hqbr.Hqbr(manga_url=comic_url, logger=logging, current_directory=current_directory,
                      sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                      chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                      keep_files=kwargs.get("keep_files"),
                      print_index=print_index)
            return 0
        elif domain in ["www.comicextra.com", "comicextra.com"]:
            comicextra.ComicExtra(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                  sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                  chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                  keep_files=kwargs.get("keep_files"),
                                  print_index=print_index)
            return 0
        #  TODO KO seems broken
        elif domain in ["www.readcomics.io", "readcomics.io"]:
            readComicsIO.ReadComicsIO(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                      sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                      chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                      keep_files=kwargs.get("keep_files"),
                                      print_index=print_index)
            return 0
        elif domain in ["www.kissmanga.com", "kissmanga.com"]:
            # kissManga.KissManga(manga_url = comic_url, logger = logging,
            #  current_directory = current_directory, sorting_order = sorting)
            print("Under Development!")
            return 0
        elif domain in ["www.bato.to", "bato.to"]:
            batoto.Batoto(manga_url=comic_url, logger=logging, current_directory=current_directory,
                          sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                          chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                          keep_files=kwargs.get("keep_files"), username=user_name, password=password,
                          comic_language=self.comic_language_resolver(comic_language),
                          print_index=print_index)
            return 0
        elif domain in ["manganelo.com", "mangakakalot.com", "manganato.com", "readmanganato.com", "chapmanganato.com"]:
            manganelo.Manganelo(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                keep_files=kwargs.get("keep_files"),
                                print_index=print_index)
            return 0
        elif domain in ["www.mangaeden.com"]:
            if print_index:
                print("please use -find and -cid instead!")
                return -1
            mangaEden.MangaEden(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                keep_files=kwargs.get("keep_files"))
            return 0
        elif domain in ["www.webtoons.com", "webtoons.com"]:
            webtoons.Webtoons(manga_url=comic_url, logger=logging, current_directory=current_directory,
                              sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                              chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                              keep_files=kwargs.get("keep_files"), image_quality=kwargs.get("image_quality"))
            return 0
        elif domain in ["www.lectortmo.com", "lectortmo.com"]:
            lectortmo.LectorTmo(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                keep_files=kwargs.get("keep_files"),
                                print_index=print_index)
        elif domain in ["www.mangatoon.mobi", "mangatoon.mobi"]:
            mangatoonMobi.MangatoonMobi(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                        sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                        chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                        keep_files=kwargs.get("keep_files"),
                                        print_index=print_index)
        elif domain in ["www.mangadex.org", "mangadex.org"]:
            mangadex.Mangadex(manga_url=comic_url, logger=logging, current_directory=current_directory,
                              sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                              chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                              keep_files=kwargs.get("keep_files"),
                              print_index=print_index)
        elif domain in ["www.qiremanhua.com", "qiremanhua.com"]:
            quiremanhua.QuireManhua(manga_url=comic_url, logger=logging, current_directory=current_directory,
                                    sorting_order=sorting, log_flag=log_flag, download_directory=download_directory,
                                    chapter_range=chapter_range, conversion=kwargs.get("conversion"),
                                    keep_files=kwargs.get("keep_files"),
                                    print_index=print_index, manual_cookies=manual_cookies)
        else:
            print("%s is not supported at the moment. You can request it on the Github repository." % domain)
